File: mobile/src/services/transactions.ts
import { apiRequest } from './api';

export interface Transaction {
  id: string;
  amount: string;
  currency: string;
  merchant: string | null;
  category_id: string | null;
  transaction_type: string;
  payment_method: string | null;
  transaction_date: string;
  notes: string | null;
  status: string;
  is_confirmed: boolean;
}

export interface TransactionList {
  items: Transaction[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateTransactionInput {
  amount: string;
  merchant?: string;
  category_id?: string;
  transaction_type: string;
  payment_method?: string;
  transaction_date: string;
  notes?: string;
}

export async function listTransactions(params?: {
  type?: string;
  page?: number;
  page_size?: number;
}): Promise<TransactionList> {
  const query = new URLSearchParams();
  if (params?.type) query.set('type', params.type);
  if (params?.page) query.set('page', String(params.page));
  if (params?.page_size) query.set('page_size', String(params.page_size));
  const qs = query.toString();
  return apiRequest<TransactionList>(`/api/v1/transactions${qs ? `?${qs}` : ''}`);
}

export async function createTransaction(data: CreateTransactionInput): Promise<Transaction> {
  return apiRequest<Transaction>('/api/v1/transactions', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function deleteTransaction(id: string): Promise<void> {
  await apiRequest<void>(`/api/v1/transactions/${id}`, { method: 'DELETE' });
}

File: backend/app/api/v1/transactions.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.user import User
from app.schemas.transaction import (
    TransactionCreate,
    TransactionListResponse,
    TransactionResponse,
    TransactionUpdate,
)
from app.services.transaction_service import (
    create_transaction,
    delete_transaction,
    get_transaction,
    list_transactions,
    update_transaction,
)

router = APIRouter(prefix="/transactions", tags=["transactions"])


@router.get("", response_model=TransactionListResponse)
def get_transactions(
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    category: UUID | None = None,
    merchant: str | None = None,
    type: str | None = Query(default=None, alias="type"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    items, total = list_transactions(
        db,
        current_user,
        date_from=date_from,
        date_to=date_to,
        category_id=category,
        merchant=merchant,
        transaction_type=type,
        page=page,
        page_size=page_size,
    )
    return TransactionListResponse(
        items=[TransactionResponse.model_validate(i) for i in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("", response_model=TransactionResponse, status_code=201)
def post_transaction(
    data: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return create_transaction(db, current_user, data)


@router.get("/{transaction_id}", response_model=TransactionResponse)
def get_transaction_by_id(
    transaction_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_transaction(db, current_user, transaction_id)


@router.patch("/{transaction_id}", response_model=TransactionResponse)
def patch_transaction(
    transaction_id: UUID,
    data: TransactionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return update_transaction(db, current_user, transaction_id, data)


@router.delete("/{transaction_id}", status_code=204)
def remove_transaction(
    transaction_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    delete_transaction(db, current_user, transaction_id)
